src/Lean/Data/Lsp/LanguageFeatures.lean

```diff
@@ -333,8 +333,8 @@ def SemanticTokenType.names : Array String :=
     "event", "method", "macro", "modifier", "comment", "string", "number",
     "regexp", "operator", "decorator", "leanSorryLike"]
 
-def SemanticTokenType.toNat (type : SemanticTokenType) : Nat :=
```

Comment:
Should this new linter be limited to situations where denoting a constructor makes sense?

Reply:
I don't think so, but it's a good question to ask.

There's two reasons:

 1. For new Lean users, it may not be clear where a constructor can or can't be denoted by a name. Having the linter fire may help make it clear that this is a variable.
 2. For less-new Lean users, avoiding this lint can make it more difficult to introduce bugs during refactoring. While I've been going through trying to make the compiler avoid this lint, there's been a few cases where renaming the argument to the function but skipping its use sites results in a program that still type checks, because the function is defined in the datatype's namespace and the constructor in question is nullary. I'm being careful while I do the renaming, and using "rename symbol", but it seems useful to have code in general that's robust in the face of incomplete renamings.

What do you think?

Reply:
There are plenty of examples in the diff of this PR where people felt like the constructor name was the most natural name to use (both because it is known to be the constructor and by accident) :-)

Reply:
> Maybe using `fvar` to bind an `Expr` known to be a `.fvar n`

This is the single most common thing that would need changing in this diff, followed by `mvar` for the same reason, and then `alt` for `Alt`.

I could add an additional setting for constructors that we don't want to warn about, or I could do an update-stage0 followed by disabling this linter in those files. Those would work fine as well - I also don't have a hugely strong opinion on this, so long as new users are getting a good experience.

Reply:
Though whether it's firing in places where constructor patterns are allowed is a bit of a red herring for this kind of situation, because we have a fair number of instances of `let fvar := ...`

Reply:
OK, I'm going to aim for less disturbance right now, and then we can have the discussion later. I'll add an option to disable it for certain constructors, set the default to false, update-stage0, then set the default to true with all of the cases that we frequently use set to exceptions. Then we can take it case by case, and not all in one ginormous diff.

Reply:
This has now been changed so that the linter only fires for nullary constructors that overlap bound variable names of the same type - only one constructor was hit by this in all of Lean.

At the same time, I also updated the error message for non-constructor things being applied in patterns so that it suggests valid names if any exist. This should also catch similar misunderstandings in those cases.

Now it doesn't need to be selectively disabled, which is a good sign.

Reply:
@david-christiansen 

>  only one constructor was hit by this in all of Lean.

I am curious what the one example was?

Reply:
Turns out I was wrong - there was this one, and then also https://github.com/leanprover/lean4/pull/4301/files#diff-ec4d9d868e0d4a5702d821884ef44382eb312b8de8fab76db01879ee97fead78L104-R108 .

So two places - not too terrible, I suppose, and I think both are easier to read with this change.